File: backend/server/src/services/tools/searchPois.js
/**
 * search_pois 工具
 * 调用高德 POI 搜索 API，搜索城市中的 POI 地点
 */

const axios = require('axios');
const config = require('../../config');

// 高德 POI 搜索 API 端点
const AMAP_POI_API = 'https://restapi.amap.com/v3/place/text';

/**
 * POI 类型映射
 * 将用户友好的类型映射到高德 API 的类型编码
 */
const POI_TYPE_MAP = {
  '景点': '风景名胜',
  '餐饮': '餐饮服务',
  '住宿': '住宿服务',
  '购物': '购物服务',
  '酒店': '住宿服务',
  '餐厅': '餐饮服务',
  '美食': '餐饮服务',
  '公园': '风景名胜',
  '博物馆': '风景名胜',
  '古迹': '风景名胜',
};

/**
 * 搜索城市中的 POI 地点
 * @param {string} city - 城市名称
 * @param {string} keywords - 搜索关键词，多个用逗号分隔
 * @param {string} type - POI 类型（可选）
 * @returns {Promise<Array>} 格式化后的 POI 列表
 */
async function searchPois(city, keywords, type = '') {
  // 构建搜索关键词
  let searchKeywords = keywords;
  if (type && POI_TYPE_MAP[type]) {
    searchKeywords = `${keywords},${POI_TYPE_MAP[type]}`;
  }

  try {
    const response = await axios.get(AMAP_POI_API, {
      params: {
        key: config.amap.webServiceKey,
        keywords: searchKeywords,
        city: city,
        citylimit: true, // 限制在指定城市内
        output: 'json',
        offset: 10,
        page: 1,
      },
      timeout: 10000, // 10秒超时
    });

    if (response.data.status !== '1' || !response.data.pois) {
      console.error('高德 POI API 返回错误:', response.data.info);
      return [];
    }

    // 格式化返回结果
    const pois = response.data.pois.map((poi, index) => ({
      id: poi.id || `poi_${index}`,
      name: poi.name,
      location: poi.location,
      address: poi.address || '',
      type: poi.type || type,
      typecode: poi.typecode || '',
      rating: parseFloat(poi.biz_ext?.rating) || 0, // 评分
      cost: poi.biz_ext?.cost || '', // 人均消费
      businessHours: poi.biz_ext?.business_hours || '', // 营业时间
    }));

    return pois;
  } catch (error) {
    console.error('POI 搜索失败:', error.message);
    throw new Error(`POI 搜索失败: ${error.message}`);
  }
}

/**
 * 根据已有 POI 列表过滤和排序
 * @param {Array} pois - POI 列表
 * @param {Object} options - 过滤选项
 * @returns {Array} 过滤后的 POI 列表
 */
function filterAndRankPois(pois, options = {}) {
  const { maxResults = 8, preferTypes = [], avoidTypes = [] } = options;

  let filtered = [...pois];

  // 排除某些类型
  if (avoidTypes.length > 0) {
    filtered = filtered.filter(poi => {
      const poiType = (poi.type || '').toLowerCase();
      return !avoidTypes.some(t => poiType.includes(t.toLowerCase()));
    });
  }

  // 优先某些类型
  if (preferTypes.length > 0) {
    filtered.sort((a, b) => {
      const aMatch = preferTypes.some(t => (a.type || '').includes(t));
      const bMatch = preferTypes.some(t => (b.type || '').includes(t));
      if (aMatch && !bMatch) return -1;
      if (!aMatch && bMatch) return 1;
      // 评分高的优先
      return (b.rating || 0) - (a.rating || 0);
    });
  } else {
    // 默认按评分排序
    filtered.sort((a, b) => (b.rating || 0) - (a.rating || 0));
  }

  // 限制返回数量
  return filtered.slice(0, maxResults);
}

module.exports = {
  searchPois,
  filterAndRankPois,
  POI_TYPE_MAP,
};

File: backend_new/src/tools/search_pois.py
import hashlib
import requests
from src.config import get_settings

settings = get_settings()


def _sign(params: dict) -> str:
    """Generate signature for Amap API request."""
    sorted_params = sorted(params.items(), key=lambda x: x[0])
    sign_str = "".join(f"{k}{v}" for k, v in sorted_params)
    sign_str += settings.AMAP_SECURITY_CODE
    return hashlib.md5(sign_str.encode()).hexdigest()


def _is_rate_limited(data) -> bool:
    """Check if response indicates rate limiting."""
    data_str = str(data).lower()
    return any(kw in data_str for kw in ["CUQPS_HAS_EXCEEDED", "rate limit", "频率", "limit exceeded"])


async def search_pois(keywords: str, city: str, types: str = None) -> list[dict]:
    """Search POIs via Amap API. Returns list of POI dicts."""
    url = "https://restapi.amap.com/v3/place/text"
    params = {
        "key": settings.AMAP_KEY,
        "keywords": keywords,
        "city": city,
        "output": "json",
        "page": 1,
        "offset": 10,
    }
    if types:
        params["types"] = types

    if settings.AMAP_SECURITY_CODE:
        params["sig"] = _sign(params)

    try:
        response = requests.get(url, params=params, timeout=10)
        data = response.json()
    except Exception:
        return []

    if _is_rate_limited(data):
        return []

    pois = data.get("pois", [])
    return [
        {
            "name": p.get("name", ""),
            "address": p.get("address", ""),
            "location": p.get("location", ""),
            "type": p.get("type", ""),
        }
        for p in pois
    ]


def filter_and_rank_pois(pois: list, options: dict = None) -> list:
    """
    根据选项过滤和排序 POI。
    对应 Node.js searchPois.js filterAndRankPois()
    """
    if options is None:
        options = {}

    max_results = options.get("max_results", 8)
    prefer_types = options.get("prefer_types", [])
    avoid_types = options.get("avoid_types", [])

    filtered = list(pois)

    # 排除某些类型
    if avoid_types:
        filtered = [
            p for p in filtered
            if not any(t.lower() in (p.get("type") or "").lower() for t in avoid_types)
        ]

    # 优先某些类型，按评分排序
    def sort_key(p):
        score = p.get("rating", 0) or 0
        if prefer_types:
            matched = any(t in (p.get("type") or "") for t in prefer_types)
            if matched:
                score += 1000
        return score

    filtered.sort(key=sort_key, reverse=True)
    return filtered[:max_results]